### PR TITLE
Fix unmarshal text on missing value.

### PR DESCRIPTION
**Pull request by alsamylkin**
We invoke `TextUnmarshaler.UnmarshalText` even if a value is missing from a config, which will lead to errors like https://github.com/uber-go/config/issues/55. We should keep a zero initialized value instead of unmarshalling an empty string.

### CODE REVIEW

File: decoder.go [hunk 1]
```diff
@@ -561,18 +570,26 @@ func (d *decoder) tryUnmarshalers(key string, value reflect.Value, def string) (
 		// Unmarshal corresponding json.
 		return true, errorWithKey(t.UnmarshalJSON(b), key)
 	case encoding.TextUnmarshaler:
+		if skip(&v, def) {
```
Comment (glibsm) [inline decoder.go]: why not do all these outside the switch?

File: decoder.go [hunk 2]
```diff
@@ -548,10 +552,15 @@ func (d *decoder) tryUnmarshalers(key string, value reflect.Value, def string) (
 	switch t := value.Addr().Interface().(type) {
 	case json.Unmarshaler:
 		// Skip unmarshaling if there is no value.
-		if !v.HasValue() {
+		if skip(&v, def) {
 			return true, nil
 		}
 
+		// check if we need to use custom defaults
```
Comment (glibsm) [inline decoder.go]: this comment means nothing to me

File: decoder.go [hunk 3]
```diff
@@ -529,6 +529,10 @@ func jsonMap(v interface{}) interface{} {
 	return v
 }
 
+func skip(value *Value, def string) bool {
```
Comment (glibsm) [inline decoder.go]: `shouldSkip`, skip implies action

File: decoder.go (hunk 2 above)
Comment (alsamylkin) [inline decoder.go]: if we don't have a value from the config provider, we should use a default.

File: decoder.go (hunk 1 above)
Comment (alsamylkin) [inline decoder.go]: because it is a type switch, and we want to return false in case if a type is not one of the unmarshalers

**Pull request merged**